### ISSUE TITLE
Extend size_limit setting in query engine to support unlimited index query.

### ISSUE THREAD

**Issue by penghuo** [opensearch-project/sql#703]
## Problem statements
Currently, the query.size_limit [setting](https://github.com/opensearch-project/sql/blob/main/docs/user/ppl/admin/settings.rst#plugins-query-size-limit) configure the maximum amount of documents to be pull from OpenSearch. The default value is: 200. for example, Let's say size_limit = 200, and index has 10K docs.
* expected, return 200 rows.
`source=index`
* expected return 1 docs
`source=index | head 1`
* return 200 rows, it is not user expected.
`source=index | head 11000`

## Proposal
The query.size_limit configure the maximum amount of rows returned by query. The default value is: 200. size_limit must larger than 0. If the query has head(PPL) or limit(SQL). it will override the query.size_limit setting.

## Expectation of search query .
* should return 200 rows.
`source=index`
* should return 1 rows.
`source=index | head 1`
* should return 11000 rows.
`source=index | head 11000`

## Expectation of aggregation query.
* should return 200 rows.
`source=index | stats request, count(*) by request`
* should return 11000 rows.              
`source=index | stats request, count(*) by request | head 11000`

**Comment by dai-chen** [comment]
Probably another way of thinking about this: the `size_limit` setting is just for default behavior. If users specify a larger number by `head` or `LIMIT`, that means they're aware of what they're doing and just want to override the default limit value. This may be safer than setting size limit to -1 and user run a query without head command later?

**Comment by penghuo** [comment]
> Probably another way of thinking about this: the `size_limit` setting is just for default behavior. If users specify a larger number by `head` or `LIMIT`, that means they're aware of what they're doing and just want to override the default limit value. This may be safer than setting size limit to -1 and user run a query without head command later?

Update the proposal as discussed.

**Comment by seankao-az** [comment]
## Design
### Physical Plan Generation
![Extend query size](https://user-images.githubusercontent.com/107227027/182191048-746b0959-8973-4153-a8e9-aeb0671fbc3a.jpg)
1. Analyzer
(1) When visiting the `Relation` node, fetch `index.max_result_window` and store it in [PlanContext](https://github.com/opensearch-project/sql/issues/656#issuecomment-1178062686)
(2) When visiting the `Limit` (SQL) or `Head` (PPL) node, decide whether scrolling is needed, and set `IndexScanType` in `PlanContext` accordingly
2. Optimizer
Use `PlanContext` to decide if scrolling is needed. If so, keep the `Limit` plan node and don't merge it with `Relation` or `IndexScan`
3. Planner
(1) For the `OpenSearchIndexScan` plan node, decide whether to use `OpenSearchQueryRequest` or `OpenSearchScrollRequest` based on `PlanContext`
(2) When pushing down `limit` and `offset`, restrict their sum to be `<= index.max_result_window`

### Execution
1. For index scan operator, fetch rows in batches of size `index.max_result_window`
2. The limit operator keeps track of the number of rows fetched

**Comment by seankao-az** [comment]
Since we're allowing query size to exceed `query.size_limit`, now there's some case we need to decide what the behavior should be.

Here we assume
```
query.size_limit = 200
index.max_result_window = 10000
```

These work as expected:
* `source=index` returns 200 rows
* `source=index | head 1` returns 1 row
* `source=index | head 300` returns 300 rows
* `source=index | head 11000` returns 11000 rows using scroll
* `source=index | fields a,b` returns 200 rows
* `source=index | fields a,b | head 1` returns 1 row

But these don't:
* `source=index | fields a,b | head 300` returns 200 rows
* `source=index | fields a,b | head 11000` returns 200 rows

The reason being that limit is only pushed down to index scan if they're optimized and merged into a single node.
In these two cases the index scan has query size 200 (query.size_limit).

### Solutions
#### Option 1

**Comment by seankao-az** [comment]
One note on the performance.
With this feature, there's no limitation on the size of the query result anymore, so it's possible that a single request-response cycle take too long and timeout.

**Comment by penghuo** [comment]
@seankao-az @dai-chen 
Want to revisit the definition of plugins.query.size_limit, 
currently, the definition of plugins.query.size_limit is , The new engine fetches a default size of index from OpenSearch set by this setting, the default value equals to max result window in index level (10000 by default). You can change the value to any value not greater than the max result window value in index level (index.max_result_window). https://github.com/opensearch-project/sql/blob/main/docs/user/admin/settings.rst#plugins-query-size-limit

In my opinion, there are two issues:
1. It is unclear how plugins.query.size_limit works.
2. It should not be tied to the max_result_window.

My proposal is 
The query.size_limit configuration sets the maximum number of rows returned by a query. The default value is 10,000, and size_limit must be greater than 0.
Note: This limit applies regardless of whether the query includes HEAD (PPL) or LIMIT (SQL).

**Comment by seankao-az** [comment]
makes sense to me. so query.size_limit can be any positive number, regardless to max_result_window.

Regarding
> It is unclear how plugins.query.size_limit works.

I think we should let plugins.query.size_limit setting only decide the final result size, not size of any intermediate step.
Currently `source=index | <other commands>`, if no other operation is pushed down to DSL, then <other commands> will operate only on the 10000 (query.size_limit) results returned from the scan.
